From 09d21cfe7372453b7954765ed45286f4982322fe Mon Sep 17 00:00:00 2001
From: Tianshu Qiu <tian.shu.qiu@intel.com>
Date: Fri, 5 Dec 2014 15:42:55 +0800
Subject: [PATCH] atomisp: Add a platform interface to remove camera sub
 device

This patch adds a new platform interface to remove
the registered camera sub device for atomisp.
This interface is called in the remove function of
camera driver.
With this patch, camera driver can support dynamically
reloading.

Change-Id: I7ad4ae7afb5a8360d7d3e5ff40d0b151a246e31d
Signed-off-by: Tianshu Qiu <tian.shu.qiu@intel.com>
---
 .../x86/platform/intel-mid/atomisp_gmin_platform.c |   29 ++++++++++++++++++++
 include/linux/atomisp_gmin_platform.h              |    1 +
 2 files changed, 30 insertions(+)

diff --git a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
index a7637bd..030037d 100644
--- a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
+++ b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
@@ -225,6 +225,35 @@ struct v4l2_subdev *atomisp_gmin_find_subdev(struct i2c_adapter *adapter,
 }
 EXPORT_SYMBOL_GPL(atomisp_gmin_find_subdev);
 
+int atomisp_gmin_remove_subdev(struct v4l2_subdev *sd)
+{
+	int i, j;
+
+	if (!sd)
+		return 0;
+
+	for (i = 0; i < MAX_SUBDEVS; i++) {
+		if (pdata.subdevs[i].subdev == sd) {
+			for (j = i + 1; j <= MAX_SUBDEVS; j++)
+				pdata.subdevs[j - 1] = pdata.subdevs[j];
+		}
+		if (gmin_subdevs[i].subdev == sd) {
+			if (gmin_subdevs[i].gpio0)
+				gpiod_put(gmin_subdevs[i].gpio0);
+			gmin_subdevs[i].gpio0 = NULL;
+			if (gmin_subdevs[i].gpio1)
+				gpiod_put(gmin_subdevs[i].gpio1);
+			gmin_subdevs[i].gpio1 = NULL;
+			if (pmic_id == PMIC_REGULATOR) {
+				regulator_put(gmin_subdevs[i].v1p8_reg);
+				regulator_put(gmin_subdevs[i].v2p8_reg);
+			}
+			gmin_subdevs[i].subdev = NULL;
+		}
+	}
+	return 0;
+}
+EXPORT_SYMBOL_GPL(atomisp_gmin_remove_subdev);
 
 struct gmin_cfg_var {
 	const char *name, *val;
diff --git a/include/linux/atomisp_gmin_platform.h b/include/linux/atomisp_gmin_platform.h
index 118c73e..24f2cbe 100644
--- a/include/linux/atomisp_gmin_platform.h
+++ b/include/linux/atomisp_gmin_platform.h
@@ -25,6 +25,7 @@ int atomisp_register_i2c_module(struct v4l2_subdev *subdev,
                                 enum intel_v4l2_subdev_type type);
 struct v4l2_subdev *atomisp_gmin_find_subdev(struct i2c_adapter *adapter,
 					     struct i2c_board_info *board_info);
+int atomisp_gmin_remove_subdev(struct v4l2_subdev *sd);
 int gmin_get_config_var(struct device *dev, const char *var, char *out, size_t *out_len);
 int gmin_get_var_int(struct device *dev, const char *var, int def);
 int camera_sensor_csi(struct v4l2_subdev *sd, u32 port,
-- 
1.7.9.5

